test/docker/presync-webhook-server/main.go

```diff
@@ -0,0 +1,170 @@
+		log.Printf("Replaced existing digest with new digest: %s", URLWithSha)
+		return URLWithSha, nil
+	}
+
+	// image URL has tag
+	URLWithSha := fmt.Sprintf("%s@sha256:%s", imageWithoutTag, commitSHA)
+	return URLWithSha, nil
+}
+
+// auth authenticates Cosign with the Docker registry using gcloud credentials.
+func auth() error {
+	authMutex.Lock()
+	defer authMutex.Unlock()
+
+	if authorized {
+		log.Printf("Skip authorizing docker as already done")
+		return nil
+	}
+	gcloudCmd := exec.Command("gcloud", "auth", "print-access-token")
+	accessTokenBytes, err := gcloudCmd.Output()
+	if err != nil {
+		return fmt.Errorf("Error fetching access token: %v\n", err)
+	}
+	accessToken := string(accessTokenBytes)
+	accessToken = accessToken[:len(accessToken)-1] // Remove the trailing newline
+
+	cmd := exec.Command("cosign", "login", "us-docker.pkg.dev", "-u", "oauth2accesstoken", "-p", accessToken)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("gcloud auth ar failed: %s, output: %s", err, string(output))
+	}
+	log.Printf("Docker authorization done: %s", string(output))
+	authorized = true
+	return nil
+}
+
+func handleWebhook(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Failed to read request body: %v", err)
+		http.Error(w, "Failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var admissionReview admissionv1.AdmissionReview
+	if err := json.Unmarshal(body, &admissionReview); err != nil {
+		log.Printf("Failed to unmarshal admission review: %v", err)
+		http.Error(w, "Failed to unmarshal admission review", http.StatusBadRequest)
+		return
+	}
+
+	oldAnnotations, err := getAnnotations(admissionReview.Request.OldObject.Raw)
+	if err != nil {
+		log.Printf("Failed to extract old annotations: %v", err)
+	}
+
+	newAnnotations, err := getAnnotations(admissionReview.Request.Object.Raw)
+	if err != nil {
+		log.Printf("Failed to extract new annotations: %v", err)
+	}
+
+	response := &admissionv1.AdmissionResponse{
+		UID: admissionReview.Request.UID,
+	}
+
+	if newAnnotations[SourceURL] != oldAnnotations[SourceURL] ||
+		newAnnotations[SourceCommit] != oldAnnotations[SourceCommit] {
+		log.Printf("Detected pre-sync annotation changes")
+		if err := auth(); err != nil {
+			log.Printf("Failed to authorize docker %v", err)
+			response.Result = &metav1.Status{
+				Message: fmt.Sprintf("Failed to authorize docker: %v", err),
+			}
+			response.Allowed = false
+		}
+		if err := validateImage(newAnnotations[SourceURL], newAnnotations[SourceCommit]); err != nil {
+			log.Printf("Image validation failed: %v", err)
+			response.Allowed = false
+			response.Result = &metav1.Status{
+				Message: fmt.Sprintf("Image validation failed: %v", err),
+			}
+		} else {
+			log.Printf("Image validation successful for %s", newAnnotations[SourceURL])
+			response.Allowed = true
+		}
+	} else {
+		log.Printf("No annotation changes detected")
```

Comment:
Modified the error message when verifying the image a bit so that it reflects which one had failed. Now the test would check for failure message for image#2 specifically when image#1 is already signed, to rule out the false positives and negatives.

Since we've agreed that the webhook would not try to verify the current syncing image on-init since there won't be any admission review request sent if the image is already synced, we don't need a local cache in the webhook server.